rs/orchestrator/src/signer.rs

```diff
@@ -27,29 +33,109 @@ impl Signer for Hsm {
         }
-        Ok(Sender::ExternalHsm {
+        Ok(Box::new(ExternalHsmSender {
             pub_key,
             sign: Arc::new(get_sign_command),
+        }))
+    }
+}
+
+/// The sender is authenticated via an external HSM device and the signature mechanism is specified
+/// through the provided function reference.
+struct ExternalHsmSender {
+    /// DER encoded public key
+    pub_key: Vec<u8>,
+    /// Function that abstracts the external HSM.
+    sign: SignBytes,
+}
+
+impl Identity for ExternalHsmSender {
+    fn sender(&self) -> Result<Principal, String> {
+        Ok(Principal::self_authenticating(self.pub_key.as_slice()))
+    }
+
+    fn public_key(&self) -> Option<Vec<u8>> {
+        Some(self.pub_key.clone())
+    }
+
+    fn sign(&self, content: &EnvelopeContent) -> Result<Signature, String> {
+        let msg = content.to_request_id().signable();
+        let signature =
+            Some((self.sign)(&msg).map_err(|err| format!("Cannot create hsm signature: {err}"))?);
+        let public_key = self.public_key();
+        Ok(Signature {
+            public_key,
+            signature,
+            delegations: None,
         })
     }
 }
 
 pub struct NodeProviderSigner {
-    keypair: Secp256k1KeyPair,
+    identity: Secp256k1Identity,
 }
 
 impl NodeProviderSigner {
     pub fn new(path: &Path) -> Option<Self> {
-        let contents = std::fs::read_to_string(path).ok()?;
-        let keypair = Secp256k1KeyPair::from_pem(&contents).ok()?;
-        Some(Self { keypair })
+        let identity = Secp256k1Identity::from_pem_file(path).ok()?;
+        Some(Self { identity })
     }
 }
 
 impl Signer for NodeProviderSigner {
-    fn get(&self) -> UtilityCommandResult<Sender> {
-        Ok(Sender::SigKeys(SigKeys::EcdsaSecp256k1(
-            self.keypair.clone(),
-        )))
+    fn get(&self) -> UtilityCommandResult<Box<dyn Identity>> {
+        Ok(Box::new(self.identity.clone()))
+    }
+}
+
+/// Signed from the node itself, with its key.
+pub struct NodeSender {
+    /// DER encoded public key
+    pub_key: Vec<u8>,
+    /// Function that signs the message id
+    sign: SignMessageId,
+}
+
+impl NodeSender {
+    pub fn new(mut pub_key: Vec<u8>, sign: SignMessageId) -> Self {
+        // The constant is the prefix of the DER encoding of the ASN.1
+        // SubjectPublicKeyInfo data structure. It can be read as follows:
+        // 0x30 0x2A: Sequence of length 42 bytes
+        //   0x30 0x05: Sequence of length 5 bytes
+        //     0x06 0x03 0x2B 0x65 0x70: OID of length 3 bytes, 1.3.101.112 (where 43 =
+        //              1 * 40 + 3)
+        //   0x03 0x21: Bit string of length 33 bytes
+        //     0x00 [raw key]: No padding [raw key]
+        let mut der_encoded: Vec<u8> = vec![
+            0x30, 0x2A, 0x30, 0x05, 0x06, 0x03, 0x2B, 0x65, 0x70, 0x03, 0x21, 0x00,
+        ];
+        der_encoded.append(&mut pub_key);
+
+        Self {
+            pub_key: der_encoded,
+            sign,
+        }
```

Comment:
```suggestion
        Self {
            pub_key,
            sign,
        }
```

The code impacted by my changes does not seem to be exhaustively tested since breaking the signing logic does not make the tests fail. For example, a simple change like this still passes the tests, even though this should break the encoding of the public key and verifications of signatures using it should fail.

This slightly lowers my confidence in whether my changes are 100% correct.

Reply:
Which tests did you run? Note that not all tests run all the time. I made the change you suggested above and then ran:
```
ict test //rs/tests/consensus/tecdsa:tecdsa_key_rotation_test
```
which failed, as expected.

Reply:
I ran the wrong tests indeed. Thank you!